src/Npgsql/Internal/PgReader.cs

```diff
@@ -597,15 +613,105 @@ async ValueTask Slow(bool async)
+            // _fieldFormat = default;
+            // _resumable = default;
+            // _fieldCompleted = default;
+
+            if (HasCurrent)
+            {
+                _currentStartPos = 0;
+                _currentBufferRequirement = default;
+                _currentSize = -1;
+                Debug.Assert(!HasCurrent);
+            }
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal ValueTask CommitAsync(bool resuming)
+    {
+        if (!Initialized)
+            return new();
+
+        if (resuming)
+        {
+            if (!Resumable)
+                ThrowHelper.ThrowInvalidOperationException("Cannot resume a non-resumable read.");
+            return new();
+        }
+
+        // We don't rely on CurrentRemaining, just to make sure we consume fully in the event of a nested scope not being disposed.
+        // Also shut down any streaming, pooled arrays etc.
+        if (_requiresCleanup || (!_fieldCompleted && FieldRemaining > 0))
+            return CommitSlow();
+
+        _fieldStartPos = -1;
+        Debug.Assert(!Initialized);
+
+        // These will always be re-initialized by Init()
+        // _fieldSize = default;
+        // _fieldFormat = default;
+        // _resumable = default;
+        // _fieldCompleted = default;
+
+        if (HasCurrent)
+        {
             _currentStartPos = 0;
             _currentBufferRequirement = default;
             _currentSize = -1;
+            Debug.Assert(!HasCurrent);
+        }
+
+        return new();
+
+        async ValueTask CommitSlow()
+        {
+            // Shut down any streaming and pooling going on on the column.
+            if (_requiresCleanup)
+            {
+                if (_userActiveStream is { IsDisposed: false })
+                    await DisposeUserActiveStream(async: true).ConfigureAwait(false);
+
+                if (_pooledArray is not null)
+                {
+                    ArrayPool.Return(_pooledArray);
+                    _pooledArray = null;
+                }
+
+                if (_charsReadReader is not null)
+                {
+                    _charsReadReader.Dispose();
+                    _charsReadReader = null;
+                    _charsRead = default;
+                }
+                _requiresCleanup = false;
+            }
+
+            await Consume(async: true, count: FieldRemaining).ConfigureAwait(false);
+
+            _fieldStartPos = -1;
             Debug.Assert(!Initialized);
+
+            // These will always be re-initialized by Init()
+            // _fieldSize = default;
+            // _fieldFormat = default;
+            // _resumable = default;
+            // _fieldCompleted = default;
+
+            if (HasCurrent)
+            {
+                _currentStartPos = 0;
+                _currentBufferRequirement = default;
+                _currentSize = -1;
+                Debug.Assert(!HasCurrent);
+            }
         }
     }
 
```

Comment:
What about adding a check to `RentArray` whether we already have a large enough array?

```csharp
_requiresCleanup = true;
var pooledArray = _pooledArray;
if (pooledArray is not null)
{
	if (pooledArray.Length >= count)
		return pooledArray;
	ArrayPool.Return(pooledArray);
}
var array = _pooledArray = ArrayPool.Rent(count);
return array;
```

Reply:
Not sure why we don't already do that yeah...

Reply:
Done